network_runner/formats.py

```diff
@@ -0,0 +1,60 @@
+import json
+from network_runner import exceptions
+
+
+def format_port_config(data, os):
+    """Use standard format json string replace port configuration string
+
+    :param data: The output of ansible play command
+    :type data: String
+    :param os: The system of the switch
+    :type os: String
+
+    :returns: String
+    """
+    # find source port configuration json string
+    source_config_json = ''
+    count = 0
+    for char in data:
+        if char == '{':
+            count += 1
+            # begin of json string
+            if count == 1:
+                source_config_json = ''
+        elif char == '}':
+            count -= 1
+            # end of json string
+            if count == 0:
+                source_config_json = source_config_json + '}'
+        if count > 0:
+            source_config_json += char
+    if count != 0:
+        raise exceptions.NetworkRunnerException(
+            'invaild json format: '.join(source_config_json))
+    # transform source port configuration to target port configuration
+    source_config = json.loads(source_config_json.replace('\n', ' '))
+    target_config = {
+        'mode': None,
+        'vlan': None,
+        # format 1-12,15,17-20
+        'trunked_vlans': '',
+    }
+    if os == "fos":
+        lines = source_config['stdout_lines'][0]
+        for line in lines:
+            # deal switchport mode
+            if line.startswith('switchport mode '):
+                target_config['mode'] = line.replace(
+                    'switchport mode ', '')
+            # deal vlan
+            if line.startswith('switchport access vlan '):
+                target_config['vlan'] = int(line.replace(
+                    'switchport access vlan ', ''))
+            # deal trunked vlans
+            if line.startswith('switchport trunk allowed vlan '):
+                target_config['trunked_vlans'] = line.replace(
+                    'switchport trunk allowed vlan ', '')
+    else:
```

Comment:
We can't force a hardcoded platfform conditional like this.
We've had trouble in the past expecting an os type to have to be added/validated for something to work.

Use case here is an enduser that wants to use a proprietary platform. They develop their ansible provider for the role and drop it into the code base on top of the upstream network-runner code. We have to make sure that end users that use this lib like that aren't going to break or need to make code modifications for it to work.

What if this "else" just dumped the raw output if it doesn't have a handler for the output?
Or if we dynamically load a function dropped into the python module and use it if its there or dump raw output otherwise?

Reply:
Thank you for your suggestion. We have the following idea: 
Add an optional flag `format` to this feature, the default is false, and the obtained content will be output directly.
If `format` set to true, the code in `formats.py` is called to process information and output. 
WDYT?

Reply:
The main purpose of this fix is to provide a common interface for third party software(OSS) that uses network-runner.

For example, if we add a function to obtain switch settings and statistical information, such as port setting obtaining, the format of values that can be obtained differ depending on the switch OS.

When third-party software(OSS) uses this feature, it does something with the retrieved values, so the third-party software(OSS) must also include switch OS-specific codes.

OSS developers may not be familiar with the Switch OS specifications, making code development and maintenance difficult.

To avoid this, I think we also need logic that returns a common value on the network-runner side. We may also want to use raw data, so it would be nice to be able to choose between a common format and raw data.

Reply:
I think we're all on the same track here. Raw is always available and the formatted output is optionally available depending on platform.

Reply:
We updated the code and added an option to get_port_conf whether to format the output. The default is false. If it is set to true, the output will be formatted according to different os.

e.g:

net_runr.get_port_conf(hostname, port)
```
TASK [network-runner : fos: get port conf] *************************************                                                                                             
ok: [testhost] => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python3"}, "changed": false, "stdout": ["!Current Configuration:\n!\ninterface  0/29\ndescrip
tion 'RX2540M4_#6_iRMC'\nswitchport mode access\nswitchport access vlan 30\nswitchport trunk allowed vlan 1-4092\nexit"], "stdout_lines": [["!Current Configuration:", "!", "
interface  0/29", "description 'RX2540M4_#6_iRMC'", "switchport mode access", "switchport access vlan 30", "switchport trunk allowed vlan 1-4092", "exit"]]}   
```

net_runr.get_port_conf(hostname, port, True)
```
TASK [network-runner : fos: get port conf] *************************************
ok: [testhost] => {"mode": "access", "vlan": 30, "trunked_vlans": "1-4092"}
```
@radez PTAL.